llvm/lib/Analysis/ConstantFolding.cpp

```diff
@@ -2764,6 +2766,33 @@ static Constant *ConstantFoldIntrinsicCall2(Intrinsic::ID IntrinsicID, Type *Ty,
                   ? *C0
                   : *C1);
 
+    case Intrinsic::scmp:
+    case Intrinsic::ucmp:
+      if (isa<PoisonValue>(Operands[0]) || isa<PoisonValue>(Operands[1]))
+        return PoisonValue::get(Ty);
+
+      if (!C0 || !C1)
+        return UndefValue::get(Ty);
+
+      if (ICmpInst::compare(*C0, *C1,
+                            IntrinsicID == Intrinsic::scmp
+                                ? ICmpInst::ICMP_SLT
+                                : ICmpInst::ICMP_ULT)) {
+        return ConstantInt::get(Ty, -1, true);
+      }
+      if (ICmpInst::compare(*C0, *C1, ICmpInst::ICMP_EQ)) {
```

Comment:
```suggestion
      if (*C0 == *C1) {
```
No need to go through compare() here.